### ISSUE TITLE
Please provide a Java wrapper

### ISSUE THREAD

**Issue by Chris2011** [f-ast/fast#1]
Hey,

is there already a java wrapper to use this in a java application? If not is it possible to provide one? I would like to use one parser for many languages.

Regards

Chris

**Comment by yijunyu** [comment]
Hello Chris,

The tool was designed as a command line one, but it will not be difficult to integrate with Java applications through system calls. I deliberately avoid implement everyone in Java mainly due to the performance considerations (the ANTLR/flatbuffers generated C/C++ libraries are more performant than the Java counterparts). Therefore, if you prefer high performance in parsing, having the wrapper is indeed a better solution. 

Can you tell me more of the functionality (API's) you prefer the Java wrapper to have? Currently, `fast` command offers a number of options, which I am not sure if they are all useful for your purposes?

Best regards,

Yijun

**Comment by Chris2011** [comment]
Hi Yijun,

thx fort he response. So my needs is to parse a file like a java file, c/cpp, js, css/less/scss etc to get the structure of the file and to check what the properties. For Java for example to see what is private, public, protected, final, static, etc. For other OO Languages too. For the others like CSS or smth similar I will have a look, what I can do with the AST.

Is this info enough for you? I don't know that much about but I want to work with the AST at the end, so that I can create nodes in a a Java Tree application to show the code (parts of it) under the file tree. https://github.com/Netbeans-Modules/NbFileStructureNode - This is my plugin, it shows me that C are classes and I are interfaces. So and when I open the class node it should show me the members inside.


Regards

Chris

**Comment by yijunyu** [comment]
This sounds a reasonable request to do. For now, the list of languages we support are 

- Java
- C/C++
- C#
- Objective C
- Python
- smali

The way I tell whether a source file is of the language is by their extension names. The AST we store are initially of XML format (thanks to SrcML and other ANTLR-based parsers we included), then we translate them into either Protobuf or flatbuffers format for efficiency. You can take a look at the "test" folder for some examples. Once you have code converted into these binary format, using the `fast` command with default options on the basis of file extension names, the information you want are all available in the tags or "kinds" of the AST tree nodes. This is for writing to binary ASTs.

For the integration with your Netbeans plugin to read these binary ASTs, there could be two alternatives:
1) Use the Java APIs to access these AST nodes, as exemplified by the example code respectively in   
    - src/gen.pb/src/main/java/com/github/gumtreediff/gen/pb/AbstractPBTreeGenerator.java
    - src/gen.fbs/src/main/java/com/github/gumtreediff/gen/fbs/AbstractFBSTreeGenerator.java
2) Use the C/C++ APIS to access these AST nodes, as exemplified by the example code in
    - src/fast-pb.cc
    - src/fast-fbs.cc

Notice that the 2nd option would be more efficient, but it requires you to provide additional arguments to the `fast` command to select the information you need, while the 1st option should be acceptable and easier to integrate with your tool. 

Hope that helps you to decide.

Yijun

**Comment by Chris2011** [comment]
So in general you say, that it makes more sense, to call the CLI from my Java with the arguments, that I need, right? Faster and we don't need any wrapper for this?

**Comment by yijunyu** [comment]
Yes, if I added the functionalities to `fast` which turns the ASTs into the format you need. But I might be quicker if you can help ;-)

**Comment by Chris2011** [comment]
Ok thx for the info, I will call the fast cli and will see what I get back.

Atm, it was only a suggestion to do a wrapper, because I have a lot of projects and this is one that I didn't work on as you can see it on the date, because of some problems of the code.

But now I thought to use a 3rd-party-parser to keep up the work.


Thx and regards

**Comment by yijunyu** [comment]
Thanks, tell me your feedback after the experiment, and we can decide what to add.
Cheers -- Yijun
